New DLC hook ModifyGeneratedUnitAppearance

Iridar opened the pull request:
Fixes #783

Initially I wanted to address this with an Event, but ran into the issue mentioned in #297; some soldiers (new rookies added to the HQ roster on the campaign start) had their appearance generated before the Event Listener had a chance to register.

So I added a DLC hook instead, which is functionally the same, because the Event Listeners would have to use ELD_Immediate in order to make any changes to the generated appearance anyway.

The hook code runs at the very end of `XGCharacterGenerator::CreateTSoldier()`. 

The Resistance Faction soldiers (Reaper, Skirmisher, Templar) in their own `CreateTSoldier()` do a `super.CreateTSoldier()`, and then modify the country and generate the nickname for the generated soldier, which would mean any changes to these aspects of the soldier appearance in the DLC hook would be disregarded.

Example of the original code:
```
function TSoldier CreateTSoldier(optional name CharacterTemplateName, optional EGender eForceGender, optional name nmCountry = '', optional int iRace = -1, optional name ArmorName)
{
	local X2SoldierClassTemplateManager ClassMgr;
	local X2SoldierClassTemplate ClassTemplate;

	kSoldier = super.CreateTSoldier('ReaperSoldier', eForceGender, nmCountry, iRace, ArmorName);
	SetCountry('Country_Reaper');
	ClassMgr = class'X2SoldierClassTemplateManager'.static.GetSoldierClassTemplateManager();
	ClassTemplate = ClassMgr.FindSoldierClassTemplate('Reaper');
	kSoldier.strNickName = GenerateNickname(ClassTemplate, kSoldier.kAppearance.iGender);

	return kSoldier;
}
```
So in order to facilitate better compatibility with the DLC hook, I have added their character generator classes to the Highlander and refactored them along these lines:
```
function TSoldier CreateTSoldier(optional name CharacterTemplateName, optional EGender eForceGender, optional name nmCountry = '', optional int iRace = -1, optional name ArmorName)
{
	kSoldier = super.CreateTSoldier('ReaperSoldier', eForceGender, nmCountry, iRace, ArmorName);
	return kSoldier;
}

function SetCountry(name nmCountry)
{
	kSoldier.nmCountry = 'Country_Reaper';
	kSoldier.kAppearance.nmFlag = kSoldier.nmCountry; // needs to be copied here for pawns -- jboswell
}

function GenerateName( int iGender, name CountryName, out string strFirst, out string strLast, optional int iRace = -1 )
{
	local X2SoldierClassTemplateManager ClassMgr;
	local X2SoldierClassTemplate ClassTemplate;

	super.GenerateName( kSoldier.kAppearance.iGender, kSoldier.nmCountry, kSoldier.strFirstName, kSoldier.strLastName, kSoldier.kAppearance.iRace );

	ClassMgr = class'X2SoldierClassTemplateManager'.static.GetSoldierClassTemplateManager();
	ClassTemplate = ClassMgr.FindSoldierClassTemplate('Reaper');
	kSoldier.strNickName = GenerateNickname(ClassTemplate, kSoldier.kAppearance.iGender);
}
```
The `SetCountry()` and `GenerateName()` functions will be called by the `super.CreateTSoldier()`. This approach is already used by the SPARK character generator class; it is weird the Firaxis didn't use the same approach for the resistance faction soldiers.

The Skirmisher character generator in particular contained a seemingly completely redundant call for `GenerateName()` which would have already been called in the exact same way by the `super.CreateTSoldier()` called a few moments before; it appears not a lot of care was given to this particular part of the code. There is a lot of potential for further refactoring, but I don't wanna fix what ain't broke without a good reason.

robojumper commented on X2WOTCCommunityHighlander/Src/XComGame/Classes/X2DownloadableContentInfo.uc:
```diff
@@ -674,4 +674,21 @@ final function string GetDisplayName()
+/// <summary>
+/// Called from XGCharacterGenerator:CreateTSoldier
+/// Has no return value, just modify the CharGen.kSoldier directly.
+/// HL-Docs: feature:ModifyGeneratedUnitAppearance; issue:783; tags:Customization
```
```suggestion
/// HL-Docs: feature:ModifyGeneratedUnitAppearance; issue:783; tags:customization
```

This may work on Windows, but fails on Unix due to case sensitive file names.

Iridar replied:
Mmf I've copied the tag exactly as it was on the documentation website, I'm guessing it gets capitalized later at some point. My version of the highlander didn't have examples of that tag used.

Iridar commented:
I've confirmed that my additions function properly with this simple example:

```
static function ModifyGeneratedUnitAppearance(XGCharacterGenerator CharGen, optional name CharacterTemplateName, optional EGender eForceGender, optional name nmCountry = '', optional int iRace = -1, optional name ArmorName, optional XComGameState_Unit UnitState, optional XComGameState UseGameState)
{
	if (CharacterTemplateName == 'Soldier')
	{
		CharGen.kSoldier.kAppearance.nmLeftArmDeco = '';
		CharGen.kSoldier.kAppearance.nmRightArmDeco = '';
		CharGen.kSoldier.kAppearance.nmLeftForearm = '';
		CharGen.kSoldier.kAppearance.nmRightForearm = '';
		CharGen.kSoldier.kAppearance.nmThighs = '';
		CharGen.kSoldier.kAppearance.nmTorsoDeco = '';
		CharGen.kSoldier.kAppearance.nmShins = '';
	}
}```
Once (if) this PR goes live, I'm gonna refactor my Clean Soldier Generation mod so it can work without Mod Class Overrides and in a more predictable and streamlined manner.

MrNiceUK commented:
#297 isn't directly applicable, any shell listeners would have already been unregistered as well, the issue was that starting rookies are generated before strategy listeners are registered.

Iridar commented:
> #297 isn't directly applicable, any shell listeners would have already been unregistered as well, the issue was that starting rookies are generated before strategy listeners are registered.

Didn't think they'd get unregistered. Then yeah.

Musashi1584 commented on X2WOTCCommunityHighlander/Src/XComGame/Classes/X2DownloadableContentInfo.uc:
```diff
@@ -674,4 +674,21 @@ final function string GetDisplayName()
+/// The UnitState and the GameState will be passed with the event only if this CreateTSoldier() function was called from CreateTSoldierFromUnit(), 
+/// which normally happens only in the shell code (TQL / Challenge Mode / Character Pool). 
+/// </summary>
+static function ModifyGeneratedUnitAppearance(XGCharacterGenerator CharGen, optional name CharacterTemplateName, optional EGender eForceGender, optional name nmCountry = '', optional int iRace = -1, optional name ArmorName, optional XComGameState_Unit UnitState, optional XComGameState UseGameState)
```
Why are all these parameters optional? especially the unit state and game state. Are there calls to the dlc hook where these are actually none?

Musashi1584 commented on X2WOTCCommunityHighlander/Src/XComGame/Classes/XGCharacterGenerator.uc:
```diff
@@ -380,6 +407,15 @@ function TSoldier CreateTSoldier( optional name CharacterTemplateName, optional
 
+	// Start issue #783
+	/// HL-Docs: ref:ModifyGeneratedUnitAppearance; issue:783
+	DLCInfos = `ONLINEEVENTMGR.GetDLCInfos(false);
```
Preferably move this to a separate method

Iridar replied:
Done.

Musashi1584 commented on X2WOTCCommunityHighlander/Src/XComGame/Classes/XGCharacterGenerator.uc:
```diff
@@ -339,6 +362,10 @@ function TSoldier CreateTSoldier( optional name CharacterTemplateName, optional
 	local TAppearance DefaultAppearance;
+	// Start variables for issue #783
+	local array<X2DownloadableContentInfo> DLCInfos;
+	local int i;
```
Nitpicking alert: not a fan of single char variable names
```suggestion
	local int Index;
```

Iridar commented:
@Musashi1584 Please review this again when you have a moment.

Iridar commented:
Sorry for the mess, deleted branch by accident.

pledbrook commented on X2WOTCCommunityHighlander/Src/XComGame/Classes/XGCharacterGenerator.uc:
```diff
@@ -380,9 +403,28 @@ function TSoldier CreateTSoldier( optional name CharacterTemplateName, optional
 }
 
+// Start issue #783
+private static function ModifyGeneratedUnitAppearance(XGCharacterGenerator CharGen, optional name CharacterTemplateName, optional EGender eForceGender, optional name nmCountry = '', optional int iRace = -1, optional name ArmorName, optional XComGameState_Unit GenerateAppearanceForUnitState, optional XComGameState GenerateAppearanceForGameState)
```
Instead of making this static and passing `self` as the argument, this should probably just be an instance function, since it is bound to an instance of `XGCharacterGenerator`.

Iridar replied:
It is done that way so that the DLC Hook gets easy access to the Character Generator class the hook is being called from, in case the hook user wants to reuse some of the methods present in that specific class.

pledbrook replied:
Yes, but you can just pass `self` to the DLC hook. If a member function requires access to `self`, then it really should be an instance function/method rather than static.

pledbrook commented on X2WOTCCommunityHighlander/Src/XComGame/Classes/XGCharacterGenerator_Reaper.uc:
```diff
@@ -0,0 +1,133 @@
+// Start issue #783
+// Normally this function calls the super.CreateTSoldier, and then manually sets the country and nickname.
+// In order to make the DLC hook for this issue more compatible with resistance faction soldiers,
+// this functionality has been moved into SetCounry() and GenerateName() methods which will be called by super.CreateTSoldier.
```
```suggestion
// this functionality has been moved into SetCountry() and GenerateName() methods which will be called by super.CreateTSoldier.
```

pledbrook commented on X2WOTCCommunityHighlander/Src/XComGame/Classes/XGCharacterGenerator_Reaper.uc:
```diff
@@ -0,0 +1,133 @@
+	kSoldier.kAppearance.nmFlag = kSoldier.nmCountry; // needs to be copied here for pawns -- jboswell
+}
+
+function GenerateName( int iGender, name CountryName, out string strFirst, out string strLast, optional int iRace = -1 )
```
```suggestion
function GenerateName(int iGender, name CountryName, out string strFirst, out string strLast, optional int iRace = -1)
```

pledbrook replied:
I see that this is just inconsistency in the original Firaxis code, so I'm not too bothered about this change.

pledbrook commented on X2WOTCCommunityHighlander/Src/XComGame/Classes/XGCharacterGenerator_Reaper.uc:
```diff
@@ -0,0 +1,133 @@
+	local X2SoldierClassTemplateManager ClassMgr;
+	local X2SoldierClassTemplate ClassTemplate;
+
+	super.GenerateName( kSoldier.kAppearance.iGender, kSoldier.nmCountry, kSoldier.strFirstName, kSoldier.strLastName, kSoldier.kAppearance.iRace );
```
```suggestion
	super.GenerateName(kSoldier.kAppearance.iGender, kSoldier.nmCountry, kSoldier.strFirstName, kSoldier.strLastName, kSoldier.kAppearance.iRace);
```

pledbrook commented on X2WOTCCommunityHighlander/Src/XComGame/Classes/XGCharacterGenerator_Templar.uc:
```diff
@@ -21,19 +21,34 @@ function X2CharacterTemplate SetCharacterTemplate(name CharacterTemplateName, na
+// Start issue #783
+// Normally this function calls the super.CreateTSoldier, and then manually sets the country and nickname.
+// In order to make the DLC hook for this issue more compatible with resistance faction soldiers,
+// this functionality has been moved into SetCounry() and GenerateName() methods which will be called by super.CreateTSoldier.
```
```suggestion
// this functionality has been moved into SetCountry() and GenerateName() methods which will be called by super.CreateTSoldier.
```

pledbrook commented on X2WOTCCommunityHighlander/Src/XComGame/Classes/XGCharacterGenerator_Reaper.uc:
```diff
@@ -26,19 +26,34 @@ function X2CharacterTemplate SetCharacterTemplate(name CharacterTemplateName, na
+	return kSoldier;
+}
+
+function SetCountry(name nmCountry)
```
I don't really understand what this change achieves. Isn't `SetCountry()` typically called after `CreateTSoldier()` anyway, overriding any values set by mods using your new hook?

Iridar replied:
I don't understand what "this change" in particular you're talking about, but the way the code is set up right now, no, the ModifyGeneratedUnitAppearance hook will run after calling SetCountry(). The vanilla Reaper CharGen did call `SetCountry()` *after* calling `CreateTSoldier()`, but with the refactored code it doesn't happen, no.

pledbrook replied:
Sorry, I wasn't very clear. "This change" refers to the refactored faction soldier character generators. And my point was that `SetCountry()` is typically called by the same code that invokes `CreateTSoldier()` in the first place.

However, I now realise that code is calling `SetCountry()` on the unit state, not the generator. My bad.

Iridar replied:
I see. Thanks for bothering with this at all.

pledbrook commented on X2WOTCCommunityHighlander/Src/XComGame/Classes/XGCharacterGenerator_Skirmisher.uc:
```diff
@@ -19,20 +19,34 @@ function X2CharacterTemplate SetCharacterTemplate(name CharacterTemplateName, na
+// Start issue #783
+// Normally this function calls the super.CreateTSoldier, and then manually sets the country and nickname.
+// In order to make the DLC hook for this issue more compatible with resistance faction soldiers,
+// this functionality has been moved into SetCounry() and GenerateName() methods which will be called by super.CreateTSoldier.
```
```suggestion
// this functionality has been moved into SetCountry() and GenerateName() methods which will be called by super.CreateTSoldier.
```

pledbrook reviewed (commented):
This seems good to me, although I don't really understand the rationale for the faction soldier refactor. It seems particularly bizarre for `SetCountry()` to completely ignore its argument and just set a hard-coded country.

pledbrook commented on X2WOTCCommunityHighlander/Src/XComGame/Classes/X2DownloadableContentInfo.uc:
```diff
@@ -674,4 +674,21 @@ final function string GetDisplayName()
\ No newline at end of file
+/// End Issue #524
+
+/// Start Issue #783
+/// <summary>
```
We don't need to remove the slashes here, but we do need to remove the one from `</summary>`, so this makes things symmetrical:
```suggestion
// Start Issue #783
// <summary>
```

pledbrook commented on X2WOTCCommunityHighlander/Src/XComGame/Classes/X2DownloadableContentInfo.uc:
```diff
@@ -674,4 +674,21 @@ final function string GetDisplayName()
+///	However, it's important *not* to call the CreateTSoldier() method, as this will retrigger the hook, potentially causing an inception loop and crashing the game.
+/// The UnitState and the GameState will be passed with the event only if this CreateTSoldier() function was called from CreateTSoldierFromUnit(), 
+/// which normally happens only in the shell code (TQL / Challenge Mode / Character Pool). 
+/// </summary>
```
We don't want this line appearing in the docs:
```suggestion
// </summary>
```

Iridar replied:
Good catch, thank you.

pledbrook commented on X2WOTCCommunityHighlander/Src/XComGame/Classes/XGCharacterGenerator.uc:
```diff
@@ -380,9 +403,28 @@ function TSoldier CreateTSoldier( optional name CharacterTemplateName, optional
 }
 
+// Start issue #783
+private static function ModifyGeneratedUnitAppearance(XGCharacterGenerator CharGen, optional name CharacterTemplateName, optional EGender eForceGender, optional name nmCountry = '', optional int iRace = -1, optional name ArmorName, optional XComGameState_Unit GenerateAppearanceForUnitState, optional XComGameState GenerateAppearanceForGameState)
+{
+	local array<X2DownloadableContentInfo> DLCInfos;
+	local int i;
+	
+	/// HL-Docs: ref:ModifyGeneratedUnitAppearance; issue:783
+	DLCInfos = `ONLINEEVENTMGR.GetDLCInfos(false);
+	for (i = 0; i < DLCInfos.Length; i++)
+	{
+		DLCInfos[i].ModifyGeneratedUnitAppearance(CharGen, CharacterTemplateName, eForceGender, nmCountry, iRace, ArmorName, GenerateAppearanceForUnitState, GenerateAppearanceForGameState);
```
Also, repeating Musashi's question, why are all the arguments optional? The only caller passes all of the arguments anyway, so you're not making uses of the `optional`. I also wonder whether it's reasonable for mods to expect all of these arguments to potentially have default/empty values.

Iridar replied:
> Yes, but you can just pass self to the DLC hook. If a member function requires access to self, then it really should be an instance function/method rather than static.

Right, make sense. Sorry it took a while. Changes made.

> why are all the arguments optional?

Because they are `optional` in the CreateTSoldier() function that calls the DLC hook, and they are not always there when CreateTSoldier() is called, for example `UITLE_SkirmishModeMenu.uc` has `Soldier = CharacterGenerator.CreateTSoldier();`.

Iridar commented:
The status shows "Changes requested", but they have already been applied. I have squashed most of the commits in this PR. If everything is in order, I'm ready to mark this "ready to merge".

robojumper commented on X2WOTCCommunityHighlander/Src/XComGame/Classes/X2DownloadableContentInfo.uc:
```diff
@@ -762,3 +762,20 @@ final function string GetDisplayName()
+/// The UnitState and the GameState will be passed to this function only if this CreateTSoldier() function was called from CreateTSoldierFromUnit(), 
+/// which normally happens only in the shell code (TQL / Challenge Mode / Character Pool). 
+// </summary>
+static function ModifyGeneratedUnitAppearance(XGCharacterGenerator CharGen, optional name CharacterTemplateName, optional EGender eForceGender, optional name nmCountry = '', optional int iRace = -1, optional name ArmorName, optional XComGameState_Unit UnitState, optional XComGameState UseGameState)
```
I still think these arguments should not be optional, considering this hook is always called with something in place of the arguments.

Iridar replied:
I'm a bit unclear on the policy here.  The "X2DLCInfo::ModifyGeneratedUnitAppearance()` is called by `XGCharacterGenerator::ModifyGeneratedUnitAppearance`, which in turn is called by `XGCharacterGenerator::CreateTSoldier()`, and all of its arguments are `optional`. I would assume that if a function is calling another function, passing its own arguments to it, the `optional` status should be inherited.

Nothing is stopping mods from calling `CreateTSoldier()` with arbitrary arguments to achieve arbitrary things, so I think marking the arguments in the "X2DLCInfo::ModifyGeneratedUnitAppearance()` as `optional` is a good way of communicating that there is a possibility of them being `none`.

robojumper replied:
I see the documentation value. My comment was motivated by the fact that `optional` only allows omitting arguments on the caller side, and we as a caller want to ensure we always pass all the arguments we have -- even if their source might be optional (like they are in `CreateTSoldier`).

Additionally, not all arguments are passed verbatim. `CreateTSoldier` does this, for example:

	if (nmCountry == '')
		nmCountry = PickOriginCountry();

so `nmCountry` is (hopefully) never empty.

robojumper commented on X2WOTCCommunityHighlander/Src/XComGame/Classes/XGCharacterGenerator_Reaper.uc:
```diff
@@ -0,0 +1,133 @@
+	return kSoldier;
+}
+
+function SetCountry(name nmCountry)
```
It would be great if the high-level idea behind these changes were described in a "compatibility" section. There are a lot of mods with custom CharacterGenerators and some info on how to make these compatible with the new hook in the same way that the existing hero ones were would be useful, especially since we *might* want to revisit this for DLC_3/Sparks.

Iridar replied:
How would I do that? Add a `compatibility` tag to HL-Docs, and add some text about "blah blah now you don't have to MCO XGCharacterGenerator"?

robojumper replied:
Not entirely. What I meant is that this hook only works for `XGCharacterGenerator` classes that explicitly received the `Set*` functions rewiring. Custom `XGCharacterGenerator` classes are in the same state as the Reaper/Skirmisher/Templar `XGCharacterGenerator` class before this PR, which would overwrite a bunch of customization properties, overriding the DLC hook.

I'd like this PR to describe the problem and the steps needed to make an `XGCharacterGenerator` work with the DLC hook in a way where it doesn't potentially overwrite DLC hook changes, both so that modders can make their `XGCharacterGenerator`s 100% compatible with the DLC hook and so that we have a reference point for the Spark `XGCharacterGenerator` if we chose to do a DLC_3 HL.

robojumper reviewed (changes requested):
Approach LGTM.

Iridar commented:
I have adjusted the documentation, removed `optional` from the arguments, and put `const` in there for most of them just in case.

robojumper reviewed (approved):
Thanks!

pledbrook commented:
@Iridar Would you mind squashing and merging this? I would do it myself, but the commit messages aren't descriptive of the change and I have no idea what to write in the squashed commit.

Iridar commented:
> @Iridar Would you mind squashing and merging this? I would do it myself, but the commit messages aren't descriptive of the change and I have no idea what to write in the squashed commit.

I squashed and rebased, but I don't have the authoritah to merge.
